BV2-211 (Fix) : 이미지 업로드 지연 개선

seongeun308 opened the pull request:
## #️⃣ Issue Number

BV2-211 

## 📝 요약(Summary)

<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
- 이미지 업로드 시간 측정 aop 추가 (수동으로 활성화)
- 이미지 리사이징 및 압축을 통해 브라우저에서 콘텐츠 다운로드 최적화
- 이미지 압축 시 jpg로 포맷을 설정해 효율적인 압축 수행
- usdz 확장자 압축 포맷 설정 제외

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [x] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 공유사항 to 리뷰어

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## 📑 API Spec

| **기능**          |                |
|-----------------|----------------|
| **HTTP Method** |                |
| **URL**         |                |

### Request Body

### Response Body


## 📸스크린샷 (선택)


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)

seongeun308 committed:
BV2-212 (Feature) : 이미지 업로드 시간 측정

BV2-212 (Feature) : 이미지 업로드 시간 측정

- aop와 micrometer를 활용
- 각각의 측정한 시간은 image.upload.time로 시작하는 log를 통해 확인
- 평균 업로드 시간은 grafana로 시각화

seongeun308 committed:
BV2-214 (fix): 이미지 리사이징 후 업로드

BV2-214 (fix): 이미지 리사이징 후 업로드

- FileDirStoragePort 인터페이스에 resizeAndStore() 추가
- 이미지 업로드 시 리사이징을 거치며 업로드
- 리사이징을 위한 thumbnailator 라이브러리 추가

seongeun308 committed:
Merge branch 'R-1.2' into fix/image-load-delay

seongeun308 committed:
BV2-214 (Fix) : 리사이징 후 화질 저하 개선

BV2-214 (Fix) : 리사이징 후 화질 저하 개선

- 리사이즈 크기 300 -> 600
- 이미지 품질 옵션 추가

seongeun308 committed:
BV2-212 (Feature) :  ImageUploadMetricsAspect 비활성화

BV2-212 (Feature) :  ImageUploadMetricsAspect 비활성화

- 설정값을 통해 aop 수동으로 실행 가능하도록 @ConditionalOnProperty 추가

seongeun308 committed:
BV2-237 (Fix) : usdz 확장자 압축 포맷 설정 제외

BV2-237 (Fix) : usdz 확장자 압축 포맷 설정 제외

- 특수 상황으로 인해 usdz 확장자는 압축 포맷 제외

shinywoon commented on file/src/main/java/file/core/aop/ImageUploadMetricsAspect.java:
```diff
@@ -0,0 +1,39 @@
+            throw new Exception(t);
+        }
+    }
+}
```
특정 어노테이션 구현 후 해당 어노테이션에 발동 조건을 걸어두면 다른 기능에서 어노테이션만 태깅하면 재활용이 가능합니다.

shinywoon commented on file/src/main/java/file/adapter/output/LocalFileStorageAdapter.java:
```diff
@@ -41,6 +42,24 @@ public Path store(MultipartFile imageFile) {
+        }
+
+        return filePath;
+    }
```
👍🏼

shinywoon commented on file/src/main/java/file/core/aop/ImageUploadMetricsAspect.java:
```diff
@@ -14,6 +15,7 @@
 @Component
 @RequiredArgsConstructor
 @Slf4j
+@ConditionalOnProperty(name = "timing.enabled", havingValue = "true")
```
자주 사용되지 않거나 특별히 사용된 어노테이션과 옵션은 주석 작성해주세요~

shinywoon commented on file/src/main/java/file/adapter/output/LocalFileStorageAdapter.java:
```diff
@@ -48,12 +48,21 @@ public Path resizeAndStore(MultipartFile imageFile) {
-                    .outputFormat(OUTPUT_FORMAT)
-                    .outputQuality(0.8)
-                    .toFile(new File(filePath.toString()));
+            if (FileUtils.getExtension(filePath.getFileName().toString()).equals(".usdz")) {
+                Thumbnails.of(imageFile.getInputStream())
+                        .size(600, 600)
+                        .outputFormat(OUTPUT_FORMAT)
+                        .outputQuality(0.8)
+                        .toFile(new File(filePath.toString()));
+            } else {
+                Thumbnails.of(imageFile.getInputStream())
+                        .size(600, 600)
+                        .outputQuality(0.8)
+                        .toFile(new File(filePath.toString()));
+            }
```
```
public class LocalFileStorageAdapter {

    private static final String OUTPUT_FORMAT = "jpg";
    private final String uploadDir;

    public LocalFileStorageAdapter(String uploadDir) {
        this.uploadDir = uploadDir;
    }

    public void storeFile(MultipartFile imageFile) {
        createDirectory(uploadDir);
        Path filePath = createFilePath(imageFile);

        try {
            if (isUsdzFile(filePath)) {
                createThumbnail(imageFile, filePath, false);
            } else {
                createThumbnail(imageFile, filePath, true);
            }
        } catch (Exception e) {
            throw new ImageStorageException(ImageErrorCode.IMAGE_UPLOAD_ERROR, e);
        }
    }

    private void createDirectory(String dir) {
        // 디렉토리 생성 로직
    }

    private Path createFilePath(MultipartFile file) {
        String fileName = StringUtils.cleanPath(Objects.requireNonNull(file.getOriginalFilename()));
        String extension = FileUtils.getExtension(fileName);

        if (!extension.equals(".usdz")) {
            extension = "." + OUTPUT_FORMAT;
        }

        String serverName = UUID.randomUUID().toString();
        return Paths.get(uploadDir, serverName + extension);
    }

    private void createThumbnail(MultipartFile imageFile, Path filePath, boolean withOutputFormat) throws IOException {
        Thumbnails.Builder<?> builder = Thumbnails.of(imageFile.getInputStream())
                .size(600, 600)
                .outputQuality(0.8);

        if (withOutputFormat) {
            builder.outputFormat(OUTPUT_FORMAT);
        }

        builder.toFile(new File(filePath.toString()));
    }

    private boolean isUsdzFile(Path filePath) {
        return FileUtils.getExtension(filePath.getFileName().toString()).equals(".usdz");
    }
}
```
Builder PTN의 경우 자기 자신(this)를 반환하기에 Chaining 이 가능합니다.
위와 같이 리팩터링 가능합니다.

shinywoon replied:
IDE 와 테스트 없이 작성된 코드이니 필요한 부분 포멧만 사용해주세요~

seongeun308 replied:
클린 코드를 생각한다면 작성해준 코드가 좋지만, 일시적으로 필요한 부분이라 코드 제거 용이성을 위해 지저분한 점 참고바랍니다!

shinywoon reviewed (commented):
좋은 코드가 많네요.
리뷰 적용 후 Merge해주세요~

seongeun308 committed:
BV2-211 (Refactor) : PR 리뷰 반영

BV2-211 (Refactor) : PR 리뷰 반영

- TrackTime 어노테이션을 활용해 aop 재활용성 높임
- @ConditionalOnProperty 관련 주석 작성
